python/ray/data/_internal/execution/legacy_compat.py

```diff
@@ -0,0 +1,223 @@
+    else:
+        output = _block_list_to_bundles(blocks, owns_blocks=owns_blocks)
+        for i in output:
+            for b in i.blocks:
+                trace_allocation(b[0], "legacy_compat.blocks_to_input_buf[1]")
+        return InputDataBuffer(output)
+
+
+def _stage_to_operator(stage: Stage, input_op: PhysicalOperator) -> PhysicalOperator:
+    """Translate a stage into a PhysicalOperator.
+
+    Args:
+        stage: The stage to translate.
+        input_op: The upstream operator (already translated).
+
+    Returns:
+        The translated operator that depends on the input data.
+    """
+
+    if isinstance(stage, OneToOneStage):
+        if stage.fn_constructor_args or stage.fn_constructor_kwargs:
+            raise NotImplementedError
+        if stage.compute != "tasks":
+            raise NotImplementedError
+
+        block_fn = stage.block_fn
+        # TODO: implement arg packing and passing for test_map_batches_extra_args
+        fn_args = (stage.fn,) if stage.fn else ()
+        fn_args = fn_args + (stage.fn_args or ())
+        fn_kwargs = stage.fn_kwargs or {}
+
+        def do_map(blocks: Iterator[Block]) -> Iterator[Block]:
+            for output_block in block_fn(blocks, *fn_args, **fn_kwargs):
+                yield output_block
+
+        return MapOperator(
+            do_map,
+            input_op,
+            name=stage.name,
+            compute_strategy=get_compute(stage.compute),
+            min_rows_per_bundle=stage.target_block_size,
+            ray_remote_args=stage.ray_remote_args,
+        )
+    elif isinstance(stage, AllToAllStage):
+        fn = stage.fn
+        block_udf = stage.block_udf
+        remote_args = stage.ray_remote_args
+        stage_name = stage.name
+
+        def bulk_fn(refs: List[RefBundle]) -> (List[RefBundle], StatsDict):
+            input_owned = all(b.owns_blocks for b in refs)
+            if isinstance(stage, RandomizeBlocksStage):
+                output_owned = input_owned  # Passthrough ownership hack.
+            else:
+                output_owned = True
+            block_list = _bundles_to_block_list(refs)
+            block_list, stats_dict = fn(block_list, input_owned, block_udf, remote_args)
+            output = _block_list_to_bundles(block_list, owns_blocks=output_owned)
+            if not stats_dict:
+                stats_dict = {stage_name: block_list.get_metadata()}
+            return output, stats_dict
+
+        return AllToAllOperator(
+            bulk_fn,
+            input_op,
+            name=stage.name,
+            num_outputs=stage.num_blocks,
+        )
+    else:
+        raise NotImplementedError
+
+
+def _bundles_to_block_list(bundles: Iterator[RefBundle]) -> BlockList:
+    blocks, metadata = [], []
+    for ref_bundle in bundles:
+        for block, meta in ref_bundle.blocks:
+            blocks.append(block)
+            metadata.append(meta)
+    owns_blocks = all(b.owns_blocks for b in bundles)
```

Comment:
I think this is logically correct with a conjunction, I wonder in what cases we will have different values of b.owns_blocks?

Reply:
I think it could happen theoretically with a union stage.